http-digest-auth/src/shellyHttpCall.ts

```diff
@@ -0,0 +1,178 @@
+/**
+ * Sample HTTP request implementation of client communication
+ * with Gen2 Shelly device with or without authorization enabled
+ * https://datatracker.ietf.org/doc/html/rfc7616
+ *
+ * Some assumptions:
+ *   Algorithm is SHA-256 (as of time of writing this is the case)
+ *   Auth name is always "admin" per design
+ */
+
+
+import http from "http";
+import crypto from "crypto";
+
+export interface authParams_t {
+  username?: string;
+  nonce: string;
+  cnonce?: string;
+  realm: string;
+  algorithm: string;
+  response?: string;
+}
+
+export interface JRPCPost_t {
+  id: number,
+  method: string;
+  params?: Record<string, unknown>;
+  auth?: authParams_t;
+}
+
+
+export const shellyHttpHashAlgo = "sha256";
+
+export function HexHash(str: string) {
+  return crypto.createHash(shellyHttpHashAlgo).update(str).digest("hex");
+};
+
+const static_noise_sha256 = ':auth:6370ec69915103833b5222b368555393393f098bfbfbb59f47e0590af135f062'; // = ':auth:'+HexHash("dummy_method:dummy_uri");
+
+
+export function isauthParams(p: any): p is authParams_t {
+  return (
+    p && typeof (p) == 'object'
+    && typeof (p.nonce) == 'string'
+    && typeof (p.realm) == 'string'
+    && typeof (p.algorithm) == 'string'
+  );
+}
+
+export function complementAuthParams(authParams: authParams_t, username: string, password: string) {
+
+  authParams.username = username;
+  authParams.nonce = String(parseInt(authParams.nonce, 16));
+  authParams.cnonce = String(Math.floor(Math.random() * 10e8));
+
+  let resp = HexHash(username + ":" + authParams.realm + ":" + password);
+  resp += ":" + authParams.nonce;
+  resp += ":1:" + authParams.cnonce + static_noise_sha256;
+
+  authParams.response = HexHash(resp);
+
+};
+
+
+export const shellyHttpUsername = "admin"; // always
```

Comment:
Do we need to export this?

Reply:
A look at this exports like a documentation. If you want to do other things with that magic string why not take it from the source. Latter if we change things and change the string you get it changed automatically plus it's just a ref that is exposed the real string inside the library can't be changed ...